### ISSUE TITLE
Threshold functions should handle problematic images gracefully

### ISSUE THREAD

**Issue by jni** [scikit-image/scikit-image#3425]
## Description
In #3402, I added a whole bunch of checks to Li thresholding so that it could handle:

- "flat" images containing only one value (return that value)
- problem images containing `np.nan` (ignore `np.nan` values in the image)
- images containing `np.inf` (ignore inf values)
- images containing *only* `np.inf` and `-np.inf` (return 0)

I think that these pathological images should get pretty much the same treatment regardless of thresholding method, so I propose to extract all that logic into a function/decorator that gets applied to all (global) thresholding methods. (Local thresholding requires a bit more thought.)

**Comment by hmaarrfk** [comment]
I like the idea of consistent behaviour.

That said, I also think that this kind of "sanitization" becomes expensive quickly.

`isnan`, `isfinite`, boolean accessing, `min`, `max`, are all `O(N)`, but for images they are really `(n^2)`. I know this pales in comparison to an `O(N^2)`, for images an `O(n^4)`, algorithm, but these things add up.

I would really be interested in ways of turning off these checks.

My experiment with [decorators](https://github.com/hmaarrfk/wabisabi) has found that adding `kwargs` is difficult to get right and requires a lot of introspection.

**Comment by jni** [comment]
@hmaarrfk Let's just say O(N) where N is the number of pixels. Took me a while to figure out what you meant by O(N^2).

Anyway, I agree, the checks are quite expensive, and what's worse, currently I do these checks multiple times (one for flatness, one for nan, one for inf...). To do this right we would need a Cython function that does a single pass through the image, and yes, a way to turn it off. We might not want to use a decorator, just manually add the kwarg (e.g. `sanitize`) to our thresholding functions.

**Comment by hmaarrfk** [comment]
@jni the way you implemented it is fine. Actually it might be the fastest way because modern computers can vectorise. 

But yea, I don’t think we should be slowing things down for a santization step that can be done once in a pipeline.

**Comment by soupault** [comment]
I actually think that we will face several way of sanitizing the input, depending on the function type (for example, in functions where the input shouldn't be flatten, just selecting the non-nan pixels doesn't make much sense).
We shall still be able to factor those checks out, but the common will become quite complex and flexible.

**Comment by jni** [comment]
@soupault, indeed, that's why I said "global", above. =) Local thresholding is a whole new kettle of fish.
